pkg/kv/kvserver/store_raft.go

```diff
@@ -269,9 +269,11 @@ func (s *Store) uncoalesceBeats(
 func (s *Store) HandleRaftRequest(
 	ctx context.Context, req *kvserverpb.RaftMessageRequest, respStream RaftMessageResponseStream,
 ) *kvpb.Error {
-	// NB: unlike the other two RaftMessageHandler methods implemented by Store,
-	// this one doesn't need to directly run through a Stopper task because it
-	// delegates all work through a raftScheduler, whose workers' lifetimes are
+	comparisonResult := s.getLocalityComparison(ctx, req.FromReplica.NodeID, req.ToReplica.NodeID)
```

Comment:
With my changes, [`TestTransferLeaseToLaggingNode`](https://github.com/cockroachdb/cockroach/blob/f7dde50216365c9f3327474068c18740f433d52f/pkg/kv/kvserver/replicate_queue_test.go#L1907) in `replicate_queue_test.go` is failing with error message `panic: use of Span after Finish. Span: storage.RaftTransport: processing batch.`. 

It appears that the issue arises from using the context in error logging within the function `HandleRaftRequest` here. I think it captured the context, but the span from the context has been finished elsewhere. I was also able to reproduce the error without my changes - by introducing a log event on master would trigger the same panic (see details below).

The failure was fixable by 
1. Passing `context.Background()` instead of the current context for error logging or
2. [Deriving](https://github.com/wenyihu6/cockroach/blob/89168bbb998c6362e063c133b3eb79b88b29bb41/pkg/kv/kvserver/replicate_queue_test.go#L1898-L1901) a new context wrapping child span for each goroutine that gets created in `delayingRaftMessageHandler`. 

But I had difficulties figuring out the root cause. I tried adding debugging statements but didn’t observe any goroutines created from [this](https://github.com/cockroachdb/cockroach/blob/7c16ac26dacfc613ab9ef32a49e903790050a984/pkg/kv/kvserver/replicate_queue_test.go#L1896-L1902) code snippet to properly finished. It also appears that this [line](https://github.com/cockroachdb/cockroach/blob/89168bbb998c6362e063c133b3eb79b88b29bb41/pkg/kv/kvserver/raft_transport.go#L505-L506) is hit frequently which should mean the stopper() function has been invoked based on the comment. But I was having a hard time to spot where the Stop() was called and which specific span has been closed. I’m also wondering if this issue is specific to this particular test case or if there is a better way to improve handling of passing context in `HandleRaftRequest`. I would really appreciate any hints or guidance on where to debug.

Way to reproduce the error: 
![Screenshot 2023-06-20 at 00.02.48.png](https://files.reviewableusercontent.io/images/reviewable/56973754/kPEfULkJSyI22d4UXvlpZG1G/Screenshot%202023-06-20%20at%2000.02.48.png)

<details><summary> Debug Output </summary> 
panic: use of Span after Finish. Span: storage.RaftTransport: processing batch. Finish previously called at: goroutine 24882 [running]:
runtime/debug.Stack()
	GOROOT/src/runtime/debug/stack.go:24 +0x64
github.com/cockroachdb/cockroach/pkg/util/tracing.(*Span).finishInternal(0x140076cdc00)
	github.com/cockroachdb/cockroach/pkg/util/tracing/span.go:256 +0x60
github.com/cockroachdb/cockroach/pkg/util/tracing.(*Span).Finish(0x14000f7b8f0?)
	github.com/cockroachdb/cockroach/pkg/util/tracing/span.go:247 +0x1c
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx.func2()
	github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:485 +0x18c
created by github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx
	github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:475 +0x358


goroutine 96155 [running]:
github.com/cockroachdb/cockroach/pkg/util/tracing.(*Span).detectUseAfterFinish(0x10cc13ae0?)
	github.com/cockroachdb/cockroach/pkg/util/tracing/span.go:182 +0x128
github.com/cockroachdb/cockroach/pkg/util/tracing.(*Span).RecordingType(0x140076cdc00)
	github.com/cockroachdb/cockroach/pkg/util/tracing/span.go:442 +0x20
github.com/cockroachdb/cockroach/pkg/util/tracing.(*Span).IsVerbose(...)
	github.com/cockroachdb/cockroach/pkg/util/tracing/span.go:450
github.com/cockroachdb/cockroach/pkg/util/log.getSpanOrEventLog({0x10945ad40, 0x14000f7b8f0})
	github.com/cockroachdb/cockroach/pkg/util/log/trace.go:92 +0x30
github.com/cockroachdb/cockroach/pkg/util/log.vEventf({0x10945ad40, 0x14000f7b8f0}, 0x0, 0x1, 0x5864aa7e?, 0x1f9?, {0x106dcaf2b, 0x10}, {0x0, 0x0, ...})
	github.com/cockroachdb/cockroach/pkg/util/log/trace.go:212 +0xb8
github.com/cockroachdb/cockroach/pkg/util/log.VEventf(...)
	github.com/cockroachdb/cockroach/pkg/util/log/trace.go:238
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Store).HandleRaftRequest(0x14002d62a80, {0x10945ad40, 0x14000f7b8f0}, 0x14002f4a420, {0x10940f3e0, 0x1400716a408})
	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/store_raft.go:272 +0x64
github.com/cockroachdb/cockroach/pkg/kv/kvserver_test.delayingRaftMessageHandler.HandleRaftRequest.func1()
	github.com/cockroachdb/cockroach/pkg/kv/kvserver_test/pkg/kv/kvserver/replicate_queue_test.go:1898 +0x88
created by github.com/cockroachdb/cockroach/pkg/kv/kvserver_test.delayingRaftMessageHandler.HandleRaftRequest
	github.com/cockroachdb/cockroach/pkg/kv/kvserver_test/pkg/kv/kvserver/replicate_queue_test.go:1896 +0x130

goroutine 1 [chan receive]:
testing.(*T).Run(0x14003aaf1e0, {0x106e2ef7e?, 0x523ece28cdb7?}, 0x1093df480)
	GOROOT/src/testing/testing.go:1494 +0x314
testing.runTests.func1(0x0?)
	GOROOT/src/testing/testing.go:1846 +0x70
testing.tRunner(0x14003aaf1e0, 0x140046bfc18)
	GOROOT/src/testing/testing.go:1446 +0x10c
testing.runTests(0x14003b36e60?, {0x10cb73240, 0x2e0, 0x2e0}, {0x14000c7b6c0?, 0x26?, 0x10cbaf920?})
	GOROOT/src/testing/testing.go:1844 +0x3f0
testing.(*M).Run(0x14003b36e60)
	GOROOT/src/testing/testing.go:1726 +0x4f0
github.com/cockroachdb/cockroach/pkg/kv/kvserver_test.TestMain(0x106e40f08?)
	github.com/cockroachdb/cockroach/pkg/kv/kvserver_test/pkg/kv/kvserver/main_test.go:35 +0xcc
main.main()
	main/bazel-out/darwin_arm64-fastbuild/bin/pkg/kv/kvserver/kvserver_test_/testmain.go:1599 +0x400

goroutine 8 [chan receive]:
github.com/cockroachdb/cockroach/pkg/util/log.flushDaemon()
	github.com/cockroachdb/cockroach/pkg/util/log/log_flush.go:75 +0x50
created by github.com/cockroachdb/cockroach/pkg/util/log.init.5
	github.com/cockroachdb/cockroach/pkg/util/log/log_flush.go:41 +0x24

goroutine 9 [chan receive]:
github.com/cockroachdb/cockroach/pkg/util/log.signalFlusher()
	github.com/cockroachdb/cockroach/pkg/util/log/log_flush.go:98 +0xb4
created by github.com/cockroachdb/cockroach/pkg/util/log.init.5
	github.com/cockroachdb/cockroach/pkg/util/log/log_flush.go:42 +0x30

goroutine 36 [syscall]:
os/signal.signal_recv()
	GOROOT/src/runtime/sigqueue.go:149 +0x2c
os/signal.loop()
	GOROOT/src/os/signal/signal_unix.go:23 +0x1c
created by os/signal.Notify.func1.1
	GOROOT/src/os/signal/signal.go:151 +0x2c

goroutine 11 [chan receive]:
github.com/cockroachdb/cockroach/pkg/util/goschedstats.init.0.func1()
	github.com/cockroachdb/cockroach/pkg/util/goschedstats/runnable.go:165 +0xb0
created by github.com/cockroachdb/cockroach/pkg/util/goschedstats.init.0
	github.com/cockroachdb/cockroach/pkg/util/goschedstats/runnable.go:157 +0x24

goroutine 15 [select]:
go.opencensus.io/stats/view.(*worker).start(0x14000a24900)
	go.opencensus.io/stats/view/external/io_opencensus_go/stats/view/worker.go:292 +0x88
created by go.opencensus.io/stats/view.init.0
	go.opencensus.io/stats/view/external/io_opencensus_go/stats/view/worker.go:34 +0xa4

goroutine 98 [sleep]:
time.Sleep(0x2faf080)
	GOROOT/src/runtime/time.go:195 +0x118
github.com/cockroachdb/cockroach/pkg/util/leaktest.AfterTest.func2()
	github.com/cockroachdb/cockroach/pkg/util/leaktest/leaktest.go:144 +0xf4
github.com/cockroachdb/cockroach/pkg/kv/kvserver_test.TestTransferLeaseToLaggingNode(0x14003aaf520)
	github.com/cockroachdb/cockroach/pkg/kv/kvserver_test/pkg/kv/kvserver/replicate_queue_test.go:2080 +0xd50
testing.tRunner(0x14003aaf520, 0x1093df480)
	GOROOT/src/testing/testing.go:1446 +0x10c
created by testing.(*T).Run
	GOROOT/src/testing/testing.go:1493 +0x300

goroutine 633 [chan receive]:
github.com/cockroachdb/cockroach/pkg/util/admission.(*tokenAllocationTicker).tick(...)
	github.com/cockroachdb/cockroach/pkg/util/admission/io_load_listener.go:334
github.com/cockroachdb/cockroach/pkg/util/admission.(*StoreGrantCoordinators).SetPebbleMetricsProvider.func1()
	github.com/cockroachdb/cockroach/pkg/util/admission/grant_coordinator.go:109 +0x88
created by github.com/cockroachdb/cockroach/pkg/util/admission.(*StoreGrantCoordinators).SetPebbleMetricsProvider
	github.com/cockroachdb/cockroach/pkg/util/admission/grant_coordinator.go:103 +0x1a8

goroutine 96446 [sleep]:
time.Sleep(0xa6e49c0)
	GOROOT/src/runtime/time.go:195 +0x118
github.com/cockroachdb/cockroach/pkg/kv/kvserver_test.delayingRaftMessageHandler.HandleRaftRequest.func1()
	github.com/cockroachdb/cockroach/pkg/kv/kvserver_test/pkg/kv/kvserver/replicate_queue_test.go:1897 +0x60
created by github.com/cockroachdb/cockroach/pkg/kv/kvserver_test.delayingRaftMessageHandler.HandleRaftRequest
	github.com/cockroachdb/cockroach/pkg/kv/kvserver_test/pkg/kv/kvserver/replicate_queue_test.go:1896 +0x130

goroutine 23842 [chan receive]:
github.com/cockroachdb/cockroach/pkg/util/admission.(*tokenAllocationTicker).tick(...)
	github.com/cockroachdb/cockroach/pkg/util/admission/io_load_listener.go:334
github.com/cockroachdb/cockroach/pkg/util/admission.(*StoreGrantCoordinators).SetPebbleMetricsProvider.func1()
	github.com/cockroachdb/cockroach/pkg/util/admission/grant_coordinator.go:109 +0x88
created by github.com/cockroachdb/cockroach/pkg/util/admission.(*StoreGrantCoordinators).SetPebbleMetricsProvider
	github.com/cockroachdb/cockroach/pkg/util/admission/grant_coordinator.go:103 +0x1a8

goroutine 96335 [sleep]:
time.Sleep(0xa6e49c0)
	GOROOT/src/runtime/time.go:195 +0x118
github.com/cockroachdb/cockroach/pkg/kv/kvserver_test.delayingRaftMessageHandler.HandleRaftRequest.func1()
	github.com/cockroachdb/cockroach/pkg/kv/kvserver_test/pkg/kv/kvserver/replicate_queue_test.go:1897 +0x60
created by github.com/cockroachdb/cockroach/pkg/kv/kvserver_test.delayingRaftMessageHandler.HandleRaftRequest
	github.com/cockroachdb/cockroach/pkg/kv/kvserver_test/pkg/kv/kvserver/replicate_queue_test.go:1896 +0x130

goroutine 96120 [sleep]:
time.Sleep(0xa6e49c0)
	GOROOT/src/runtime/time.go:195 +0x118
github.com/cockroachdb/cockroach/pkg/kv/kvserver_test.delayingRaftMessageHandler.HandleRaftRequest.func1()
	github.com/cockroachdb/cockroach/pkg/kv/kvserver_test/pkg/kv/kvserver/replicate_queue_test.go:1897 +0x60
created by github.com/cockroachdb/cockroach/pkg/kv/kvserver_test.delayingRaftMessageHandler.HandleRaftRequest
	github.com/cockroachdb/cockroach/pkg/kv/kvserver_test/pkg/kv/kvserver/replicate_queue_test.go:1896 +0x130
I230620 05:02:33.400240 1 (gostd) testmain.go:1570  [-] 1  Test //pkg/kv/kvserver:kvserver_test exited with error code 2
FAIL: //pkg/kv/kvserver:kvserver_test (see /private/var/tmp/_bazel_wenyihu/e6cc8641598a1deef0c08146e7d85e4b/execroot/com_github_cockroachdb_cockroach/bazel-out/darwin_arm64-fastbuild/testlogs/pkg/kv/kvserver/kvserver_test/test.log)
INFO: Elapsed time: 41.238s, Critical Path: 19.70s
INFO: 4 processes: 1 internal, 3 darwin-sandbox.
//pkg/kv/kvserver:kvserver_test                                          FAILED in 7.1s
  /private/var/tmp/_bazel_wenyihu/e6cc8641598a1deef0c08146e7d85e4b/execroot/com_github_cockroachdb_cockroach/bazel-out/darwin_arm64-fastbuild/testlogs/pkg/kv/kvserver/kvserver_test/test.log

Executed 1 out of 1 test: 1 fails locally.
INFO: Build Event Protocol files produced successfully.
INFO: Build completed, 1 test FAILED, 4 total actions
ERROR: exit status 3